--- a/thirdparty/stb_image.h
+++ b/thirdparty/stb_image.h
@@ -620,7 +620,7 @@
    #if (defined(__GNUC__) && (__GNUC__ >= 4)) || defined(__clang__)
       #define STBI_THREAD_LOCAL       __thread
    #elif defined(_MSC_VER)
       #define STBI_THREAD_LOCAL       __declspec(thread)
    #elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_THREADS__)
-      #define STBI_THREAD_LOCAL       thread_local
+      #define STBI_THREAD_LOCAL       
    #endif
 
    #ifndef STBI_THREAD_LOCAL
